data_utils/ds_addmult_gen/ds-addmult-let.py

```diff
@@ -0,0 +1,191 @@
+        if op == '+':
+            res = (e1 + e2) % 10
+            return (res, '__'+e1s+e2s+str(res))
+        elif op == '*':
+            res = (e1 * e2) % 10
+            return (res, '__'+e1s+e2s+str(res))
+        raise Exception("Unknown operator")
+
+    def let(self, tree):
+        defn, expr = tree.children
+
+        self.state.append({}) # new closure
+        _, s_defn = self.visit(defn)
+        result, s_expr = self.visit(expr)
+        self.state.pop() # exit closure
+        return (result, '__'+s_defn+s_expr+str(result))
+    
+    def defn(self, tree):
+        var, expr = tree.children
+        var = var.children[0].value
+        val, s_exp = self.visit(expr)
+
+        self.state[-1][var] = val
+        return (None, '__'+s_exp+str(val))
+
+    def var(self, tree):
+        var = tree.children[0].value
+        return (self._getvar(var), '_')
+
+    def add(self, tree):
+        return '+'
+    def mul(self,tree):
+        return '*'
+    def digit(self, tree):
+        n = int(tree.children[0].value)
+        return (n, '_')
+
+parser = lark.Lark(calc_grammar, start='expr')
+interpreter = AddMultLetInterpreter()
+
+testexp = "(L[x4](+x2))" # x=4, x+2
+testexp = "(L [x4] (L [y(+1(*2x))] (*y3)) )" # x=4, y=2x+1, 3y
+
+testexp = testexp.replace(" ", "") # whitespace is for humans
+testtree = parser.parse(testexp)
+testans = interpreter.visit(testtree)
+print(testans)
+
+
+def get_height(tree):
+    "Get the height of a Lark tree (height 1 = single digit)"
+    if hasattr(tree, 'children') and len(tree.children) > 0:
+        return 1 + max([get_height(c) for c in tree.children])
+    else:
+        return 0
+
+def make_data(out_file, n_examples, min_height, max_height, max_ans):
+    """
+    Generate problems based on the grammar, and write them to a CSV file.
+    - out_file: path to the output CSV file
+    - n_examples: how many examples to generate in total
+    - min_height: minimum height of the generated examples
+    - max_height: maximum height of the generated examples
+    - exclude_strings: if not None, separate out examples that contain one of these strings
+    - exclude_file: Write the excluded examples to this file
+    """
+
+    parser = lark.Lark(calc_grammar, start='expr')
+    interpreter = AddMultLetInterpreter()
+    generator = from_lark(parser, start="expr")
+
+    n_good = 0
+    with open(out_file, 'w') as f:
+        csv_writer = csv.DictWriter(f, fieldnames=['height', 'width', 'example', 'answer', 'ans_mod10', 'ans_sublabels', 'tree_sig'])
+        csv_writer.writeheader()
+
+        while True:
+            interpreter.reset()
+            example = generator.example()
+            tree = parser.parse(example)
+            height = get_height(tree)
+            if height < min_height or height > max_height:
+                continue
+            else:
+                try:
+                    answer, ans_inter = interpreter.visit(parser.parse(example))
+                except:
```

Comment:
Hmm is there a strong reason to do this for non-prod code?

Seems like unnecessary overhead to dive into the third party library and dig up all the possible exceptions it can raise while parsing or tree walking. e.g. there might not be a single parent parser exception.

I recognize the intention in prod to avoid mishandling OOM, disk full, IO errors, but here they'd just bubble upwards outside the try anyway?

Reply:
TLDR: I would argue yes, because bare except / continue will not always bubble upwards outside the try in obvious ways, and will make debug time heavier. I would always recommend it when more than one person is collaborating on code, too.

---

IMO this is the minimum required for non-prod code, if there is no specific exception, then just ensure the program breaks.
```
try:
# statement
except:
      raise AssertionError("One of many possible problems happened")
```

However letting the function continue regardless of the internal error is dangerous because something unintended can happen and continue for the entire duration of the program.
```
try:
# statement
except:
      continue
```
In that case I would just recommend to let the statement run so that python can handle the error normally.
```
# statement
```

At the minimum, with `except: continue` it may OOM / disk / IO like you mention, but the fear is that once it is built upon (nested in another file) it will be non-obvious to find and debug this specific issue. It sort of gets into the area where only a profiler can help fix (much heavier debug time, even for non-prod + individual code).

This particular file in reality would be okay, it doesn't seem like many people will contribute massively on top of it. Still, some tricky things could happen, like it will except `KeyboardInterrupt`, or certain useful expressions generated from this call might not get appended to our csv.

---
My suggestion on fixing quickly:

1. Check if you need try/except at all in the first place by just running the statement itself. If it is not expected to break then just keep it without try/except.
```
# statement
```
3. If it *does* break without the try/except, then copy and paste the error that it has into the except statement.

```
# statement <- raises ValueError
```
then
```
try:
# statement
except ValueError as e:
      print(f"caught ValueError: {e}")
      continue
```


Finally, I find that it is super rare to find another error inside, but if there are two potential errors or more, it is probably good to note these and log the cases, so no documentation is generally required to fix this (you just have to let the code break once or twice). 

Most devs should mark which type of error their function will raise, but I agree in this case it is tedious compared to just letting the software break itself (and devs dont always point out what will be raised in docs).

Reply:
Great point about what happens when people build upon it. I'd be super annoyed to find this as the thing eating bugs in the heart of some library I was using!

That said I do know this is a weird code smell, and I wouldn't use `except: continue` unless there was some reason for it. I don't remember what it was. So if I have it break execution, it'll probably break our code. To balance, what I'll do is have it continue but print the exception every time.

Reply:
That way we'll get visibility on what this was used for, as well.